test/e2e/app-dir/next-after/index.test.ts

```diff
@@ -0,0 +1,146 @@
+  })
+
+  if (isNextDev) {
+    it('errors with `dynamic = "error"`', async () => {
+      const filePath = 'app/static/page.js'
+      const origContent = await next.readFile(filePath)
+
+      try {
+        await next.patchFile(filePath, (contents) =>
+          contents.replace(
+            `export const dynamic = 'force-static'`,
+            `export const dynamic = 'error'`
+          )
+        )
+        const browser = await next.browser('/static')
+
+        expect(await hasRedbox(browser)).toBe(true)
+        console.log(await getRedboxDescription(browser))
+        expect(await getRedboxDescription(browser)).toContain(
+          'Route /static with `dynamic = "error"` couldn\'t be rendered statically because it used `unstable_after`'
+        )
+        expect(Log.readPersistentLog(logFile)).toHaveLength(0)
+      } finally {
+        await next.patchFile(filePath, origContent)
+      }
+    })
+  }
+
+  describe('interrupted RSC renders', () => {
+    // TODO(after): if we don't want after() to run if a redirect/notFound/error occurred,
+    // we need a way to know that it happened... somehow
+
+    it.failing('does not run callbacks if redirect() was called', async () => {
+      await next.browser('/interrupted/calls-redirect')
+      expect(Log.readPersistentLog(logFile)).not.toContainEqual({
+        source: '[page] /interrupted/calls-redirect',
+      })
+      expect(Log.readPersistentLog(logFile)).toContainEqual({
+        source: '[page] /interrupted/redirect-target',
+      })
+    })
+
+    it.failing('does not run callbacks if notFound() was called', async () => {
+      await next.browser('/interrupted/calls-not-found')
+      expect(Log.readPersistentLog(logFile)).not.toContainEqual({
+        source: '[page] /interrupted/calls-not-found',
+      })
+    })
+
+    it.failing(
+      'does not run callbacks if a user error was thrown in the RSC render',
+      async () => {
+        await next.browser('/interrupted/throws-error')
+        expect(Log.readPersistentLog(logFile)).not.toContainEqual({
+          source: '[page] /interrupted/throws-error',
+        })
+      }
+    )
+  })
+
+  it('runs in middleware', async () => {
+    const requestId = `${Date.now()}`
+    const res = await next.fetch(
+      `/middleware/redirect-source?requestId=${requestId}`,
+      {
+        redirect: 'follow',
+        headers: {
+          cookie: 'testCookie=testValue',
+        },
+      }
+    )
+
+    expect(res.status).toBe(200)
+    const cliLogs = Log.readCliLogs(next.cliOutput)
+    expect(cliLogs).toContainEqual({
+      source: '[middleware] /middleware/redirect',
+      requestId,
+      cookies: { testCookie: 'testValue' },
+    })
+  })
+
+  it.todo('runs in getMetadata()')
+  it.todo('runs after the response is sent')
+  it.todo('does not allow modifying cookies')
+  it.todo('errors when used in client modules')
+  it.todo('errors when used in pages dir')
+})
```

Comment:
Since it looks like you can call `after()` multiple times within the same render, do you need a test for that case?

Reply:
already checking that indirectly e.g. here
https://github.com/vercel/next.js/pull/65038/files#diff-9286c4c902fede24808c08467817f9c286ceab382da1fbbfe8729ff176d55494R25-R29

Reply:
We'd ideally be calling it every time we check an experiment and then deduplicating posting logs at the end.

Reply:
Ah, great!